### PR TITLE
Implement update sink

### PR DESCRIPTION
**Pull request by KochTobi**
Implements the sample update sink

### CODE REVIEW

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/UserService.groovy [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * @since 1.0.0
+ */
+interface UserService {
+    /**
+     * Retrieves user details for a provided userId from the persistence layer.
+     * @param userId the identifier of the user
+     * @return user details for a user found.
+     * @since 1.0.0
+     */
+    UserDetails getUserDetails(String userId)
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/UserService.groovy]: What happens if no user details can be retrieved for a given id? I think the API can profit from returning an `Optional<UserDetails>`.

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/ConnectionProvider.groovy [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.sql.Connection
+import java.sql.SQLException
+
+/**
+ * Provides the ability to connect to a SQL ressource
+ *
+ * @since: 1.0.0
+ */
+interface ConnectionProvider {
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/ConnectionProvider.groovy]: I think we should go for Hibernate here, instead of hazzling with the connection and queries manually

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import java.sql.SQLException
+
+/**
+ * Creates a connection to the user database
+ *
+ * A class for setting up the connection to the user database. It should be used when data needs to be retrieved from the
+ * DB or written into it.
+ *
+ * @since: 1.0.0
+ * @author: Jennifer Bödker
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy:22]: Author tag?

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import java.sql.Connection
+import java.sql.SQLException
+
+/**
+ * Creates a connection to the user database
+ *
+ * A class for setting up the connection to the user database. It should be used when data needs to be retrieved from the
+ * DB or written into it.
+ *
+ * @since: 1.0.0
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy:21]: wrong version

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+@Singleton(lazy = true)
+@Component
+class DatabaseSession implements ConnectionProvider {
+
+    private static DatabaseSession INSTANCE
+
+    private BasicDataSource dataSource
+
+    @Autowired
+    private UserDatabaseConfig userDatabaseConfig
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy:30]: I think the abstraction is a little bit overkill. I dont see, where we have different config implementations.
Imo it is sufficient to ingest the configuration values directly in the service implementation (constructor ingestion)

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+     * @param password the password to use for the database connection
+     * @param host the database host
+     * @param port the port on which the database is hosted
+     * @param sqlDatabase the name of the database
+     */
+    void init(String user,
+              String password,
+              String host,
+              String port,
+              String sqlDatabase) {
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy]: you can ingest the config values here directly with the `@Value` annotation

File: src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 7]
```diff
@@ -0,0 +1,53 @@
+{
```
Comment (sven1103) [inline src/main/resources/META-INF/additional-spring-configuration-metadata.json:1]: nice, can you tell me more about the purpose of this metadata file?

File: src/main/groovy/life/qbic/samplestatus/reporter/api/UserDetails.groovy [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    UserDetails(String firstName, String lastName, String email) {
+        this.firstName = firstName
+        this.lastName = lastName
+        this.email = email
+    }
+
+    String getFullName() {
+        return getFirstName() + " " + getLastName()
+    }
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/api/UserDetails.groovy]: We should go for JPA entities here, since we retrieve data from our persistence layer and can avoid boiler plate code of mapping pure data back and forth

File: src/main/groovy/life/qbic/samplestatus/reporter/services/NcctLocationService.groovy [hunk 9]
```diff
@@ -23,8 +25,17 @@ class NcctLocationService implements LocationService {
     @Autowired
     private SampleTrackingService sampleTrackingService
 
+    @Autowired
+    private UserService userService
+
     @Override
-    Optional<Location> getCurrentLocation() {
+    Location getCurrentLocation() {
```
Comment (sven1103) [inline src/main/groovy/life/qbic/samplestatus/reporter/services/NcctLocationService.groovy]: Why was this removed? The Optional facilitates idiomaic usage of safe value access and avoid NPEs.

**Review by sven1103 — CHANGES_REQUESTED** [review]
Only minor requests and comments, I will happily contribute to add the Hibernate implementation

File: src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy (hunk 3 above)
Comment (KochTobi) [reply src/main/groovy/life/qbic/samplestatus/reporter/services/users/database/DatabaseSession.groovy:22]: CPP - copy paste programming

File: src/main/groovy/life/qbic/samplestatus/reporter/api/UserDetails.groovy (hunk 8 above)
Comment (KochTobi) [reply src/main/groovy/life/qbic/samplestatus/reporter/api/UserDetails.groovy]: The thing is that the database for this data is currently convoluted with a lot of other fields. I do not know how to handle this case. To make it work, I implemented it this way first.

File: src/main/resources/META-INF/additional-spring-configuration-metadata.json (hunk 7 above)
Comment (KochTobi) [reply src/main/resources/META-INF/additional-spring-configuration-metadata.json:1]: https://docs.spring.io/spring-boot/docs/current/reference/html/configuration-metadata.html
IntelliJ did this for me, I think we can remove them for now and add them in a more helpful condition later.

File: src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy [hunk 10]
```diff
@@ -18,7 +18,11 @@ interface SampleTrackingService {
      * @return the location that this user is associated with
      * @since 1.0.0
      */
+<<<<<<< HEAD:src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy]:
```suggestion
```

File: src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy [hunk 11]
```diff
@@ -18,7 +18,11 @@ interface SampleTrackingService {
...
      Optional<Location> getLocationForUser(String userId)
+=======
+    Optional<Location> getLocationForUser(String userId)
+>>>>>>> 19690c18aee5fc51802bb21a2df0f89a9a046d6e:src/main/groovy/life/qbic/samplestatus/reporter/services/SampleTrackingService.groovy
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/samplestatus/reporter/api/SampleTrackingService.groovy]:
```suggestion
```

**Review by KochTobi — COMMENTED** [review]
I found a merge conflict 😄

File: src/main/resources/hibernate.cfg.xml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+<hibernate-configuration>
+
+    <session-factory>
+
+        <!-- JDBC Database connection settings -->
+        <property name="connection.driver_class">com.mysql.cj.jdbc.Driver</property>
+        <property name="connection.url">jdbc:mysql://134.2.24.116:3306/qbic_usermanagement_db?useSSL=false&amp;serverTimezone=UTC</property>
+        <property name="connection.username">qbicuser</property>
+        <property name="connection.password">*G@#588Naq1yrkSc1d2RL2Pt5a%wgr8S</property>
```
Comment (sven1103) [inline src/main/resources/hibernate.cfg.xml]: remind to the future me: use f* env vars!

Comment (KochTobi) [reply src/main/resources/hibernate.cfg.xml]: will do :+1:

**Review by sven1103 — APPROVED** [review]
